Sell/FulfilmentPoint/Repository/FulfilmentPointRepository.swift

```diff
@@ -0,0 +1,68 @@
+//
+//  FulfilmentPointRepository.swift
+//  RealifeTech
+//
+//  Created by Mickey Lee on 21/06/2021.
+//  Copyright © 2021 Realife Tech. All rights reserved.
+//
+
+import Foundation
+
+public final class FulfilmentPointRepository {
+
+    private let graphQLManager: GraphQLManageable
+
+    init(graphQLManager: GraphQLManageable) {
+        self.graphQLManager = graphQLManager
+    }
+}
+
+extension FulfilmentPointRepository: FulfilmentPointProvidable {
+
+    public func getFulfilmentPointCategories(
+        pageSize: Int,
+        page: Int,
+        callback: @escaping (Result<PaginatedObject<FulfilmentPointCategory>, Error>) -> Void
+    ) {
+        graphQLManager.dispatch(
+            query: ApolloType.GetFulfilmentPointCategoriesQuery(pageSize: pageSize, page: page),
+            cachePolicy: .returnCacheDataAndFetch
+        ) { result in
+            switch result {
+            case .success(let response):
+                let items = response.data?.getFulfilmentPointCategories?.edges?.compactMap {
```

Comment:
I'm thinking about the upcoming error handling we're going to have to deal with & that under the success case, we'll have to check for our server's errors, and possibly we should be doing that/thinking about a pattern now.

- [ ] let me know what you think

Reply:
I have the same concern about having error handling in the success case. I can check if the `data` is existed in the `GraphQLResult` in GraphQLManager.
```
private func processGraphQLResult<Data: GraphQLSelectionSet>(
        _ result: Result<GraphQLResult<Data>, Error>,
        completion: @escaping (Result<Data, Error>) -> Void
    ) {
        switch result {
        case .success(let response):
            guard
                let data = response.data,
                !data.jsonObject.isEmpty
            else {
                return completion(.failure(GraphQLError.noDataError))
            }
            completion(.success(data))
        case .failure(let error):
            completion(.failure(error))
        }
    }
```

So the callback of the GraphQLManageable would be:
```
public func dispatch<Query: GraphQLQuery>(
        query: Query,
        cachePolicy: GraphNetworkCachePolicy,
        completion: @escaping (Result<Query.Data, Error>) -> Void
    )
```

However, we still have to safe unwrap the returned Data as the query/mutation type like `getFulfilmentPointCategory` is optional in Apollo SDK and I couldn't find a generic way to safe unwrap the fragments and the normal type. That's the reason why GraphQL is not friendly to FE from my opinion.
```
guard
    let fulfilmentPointCategory = FulfilmentPointCategory(
        response: response.getFulfilmentPointCategory?.fragments.fragmentFulfilmentPointCategory)
else {
    return callback(.failure(GraphQLError.noDataError))
}
```

Anyway, do you think we need to implement this? We convert the result type to `Result<Query.Data, Error>` in GraphQLManager, so that it will throw the empty error when there's no data to map first. And in each repository layer, we still have to safe unwrap the response and throw the empty error under the success case. Really redundant solution honestly. Welcome any more generic solution

Reply:
I thought about using the Manager to handle some of this, but I don't think it matches the way GraphQL behaves. Reason being, we can get data AND have an error, or multiple errors, in the success case. It looks something like:

--> Get response, wrap in a result
       --> [failure] We have a single non-GraphQL error --> do normal error handling
       --> [success] We need to check if we have errors and if we have data
             --> [Error and no data] Do error handling as normal?
             --> [Multiple errors and no data] ???
             --> [Errors AND some data] ???
             
Basically, having errors and having data aren't exclusive states in GraphQL, so we can't just catch errors in the successful response and map them back into a Result type. Maybe we could do something like this, but I think there must be something more elegant:

```swift
enum GraphQLResponse {
    case totalSuccess(Query.Data)
    case partialSuccess(Query.Data, [GraphQLError])
    case totalFailure([GraphQLError])
    case serverNeverReached(Error) // This is the current result.failure state
}
```